README.rst
MIDIUtil
========

|build|

This is just a brief adumbration. Full documentation for the release
version can be found at `Read the Docs
<http://midiutil.readthedocs.io/>`_. The documentation for the
development version is `here <http://midiutil.readthedocs.io/en/latest/>`_.

|docs|

Introduction
------------

MIDIUtil is a pure Python library that allows one to write multi-track
Musical Instrument Digital Interface (MIDI) files from within Python
programs (both format 1 and format 2 files are now supported).
It is object-oriented and allows one to create and write these
files with a minimum of fuss.

MIDIUtil isn't a full implementation of the MIDI specification. The actual
specification is a large, sprawling document which has organically grown
over the course of decades. I have selectively implemented some of the
more useful and common aspects of the specification. The choices have
been somewhat idiosyncratic; I largely implemented what I needed. When
I decided that it could be of use to other people I fleshed it out a bit,
but there are still things missing. Regardless, the code is fairly easy to
understand and well structured. Additions can be made to the library by
anyone with a good working knowledge of the MIDI file format and a good,
working knowledge of Python. Documentation for extending the library
is provided.

This software was originally developed with Python 2.5.2 and made use of
some features that were introduced in 2.5. More recently Python 2 and 3
support has been unified, so the code should work in both environments.
However, support for versions of Python previous to 2.7 has been dropped.
Any mission-critical music generation systems should probably be updated
to a version of Python supported and maintained by the Python foundation,
lest society devolve into lawlessness.

This software is distributed under an Open Source license and you are
free to use it as you see fit, provided that attribution is maintained.
See License.txt in the source distribution for details.

Installation
------------

The latest, stable version of MIDIUtil is hosted at the `Python Package
Index <https://pypi.python.org/pypi/MIDIUtil/>`__ and can be installed
via the normal channels:

.. code:: bash

  pip install MIDIUtil

Source code is available on `Github <https://github.com/MarkCWirt/MIDIUtil>`__ ,
and be cloned with one of the following URLS:

.. code:: bash

    git clone git@github.com:MarkCWirt/MIDIUtil.git
    # or
    git clone https://github.com/MarkCWirt/MIDIUtil.git

depending on if you want to use SSH or HTTPS. (The source code
for stable releases can also be downloaded from the
`Releases <https://github.com/MarkCWirt/MIDIUtil/releases>`__
page.)

To use the library one can either install it on one's system:

.. code:: bash

    python setup.py install

or point your ``$PYTHONPATH`` environment variable to the directory
containing ``midiutil`` (i.e., ``src``).

MIDIUtil is pure Python and should work on any platform to which
Python has been ported.

If you're using this software in your own projects
you may want to consider distributing the library bundled with yours;
the library is small and self-contained, and such bundling makes things
more convenient for your users. The best way of doing this is probably
to copy the midiutil directory directly to your package directory and
then refer to it with a fully qualified name. This will prevent it from
conflicting with any version of the software that may be installed on
the target system.


Quick Start
-----------

Using the software is easy:

* The package must be imported into your namespace
* A MIDIFile object is created
* Events (notes, tempo-changes, etc.) are added to the object
* The MIDI file is written to disk.

Detailed documentation is provided; what follows is a simple example
to get you going quickly. In this example we'll create a one track MIDI
File, assign a tempo to the track, and write a C-Major scale. Then we
write it to disk.

.. code:: python

    #!/usr/bin/env python

    from midiutil import MIDIFile

    degrees  = [60, 62, 64, 65, 67, 69, 71, 72]  # MIDI note number
    track    = 0
    channel  = 0
    time     = 0    # In beats
    duration = 1    # In beats
    tempo    = 60   # In BPM
    volume   = 100  # 0-127, as per the MIDI standard

    MyMIDI = MIDIFile(1)  # One track, defaults to format 1 (tempo track is created
                          # automatically)
    MyMIDI.addTempo(track, time, tempo)

    for i, pitch in enumerate(degrees):
        MyMIDI.addNote(track, channel, pitch, time + i, duration, volume)

    with open("major-scale.mid", "wb") as output_file:
        MyMIDI.writeFile(output_file)

There are several additional event types that can be added and there are
various options available for creating the MIDIFile object, but the above
is sufficient to begin using the library and creating note sequences.

The above code is found in machine-readable form in the examples directory.
A detailed class reference and documentation describing how to extend
the library is provided in the documentation directory.

Have fun!

Thank You
---------

I'd like to mention the following people who have given feedback, bug
fixes,  and suggestions on the library:

* Bram de Jong
* Mike Reeves-McMillan
* Egg Syntax
* Nils Gey
* Francis G.
* cclauss (Code formating cleanup and PEP-8 stuff, which I'm not good at following).
* Philippe-Adrien Nousse (Adphi) for the pitch bend implementation.
* meteorsw (https://github.com/meteorsw) for major restructuring and clean-up
  of code.

I've actually been off email for a few years, so I'm sure there are lots
of suggestions waiting. Stay tuned for updates and bug fixes!

.. |docs| image:: https://readthedocs.org/projects/midiutil/badge/?version=1.1.3
   :target: http://midiutil.readthedocs.io/
   :alt: Documentation Status

.. |build| image:: https://travis-ci.org/MarkCWirt/MIDIUtil.svg?branch=master
   :target: https://travis-ci.org/MarkCWirt/MIDIUtil


CHANGELOG
Date:       4 March 2018
Version:    1.2.1

    * Fairly major restructuring of the code contributed by meteorsw, including:
      * Removing the MIDIEvent class and subsuming functionality in GenericEvent
        and derivative classes.
      * Allowing for the specification of times in MIDI ticks, as well as quarter
        notes.
      * General clean-up and restructuring. Logic on how an event serializes itself
        to the MIDI stream is now associated directly with the object.
      * Renaming of some variable to be less confusing to users.
      * Placing comparison operators in the derived classes of GenericEvent, where
        they really belong.
      * Addition of channel pressure event.
    * Clarifications on documentation.
    * Added pitch bend support with `addPitchWheelEvent`.

Date:       6 March 2017
Version:    1.1.3

    A code clean-up release. Not much in the way of new features.

    * Documentation changes, typo corrections, etc.
    * Simplify the importation of the library. It's now possible to just:

        from midiutil import MIDIFile

      which seems somewhat easier.
    * Some formatting and cleanup work (PEP-ish stuff).
    * Added Python 3.6 testing/support.
    * Removed support for Python 2.6. All good things eventually come to an
      end

Date:       28 September 2016
Version:    1.1.1

    * Just a couple of typo's corrected (what's a "but fix" anyway?),
      and the display of README.rst on Pypi.

Version:    1.1.0
    * Allow for the specification of either MIDI format 1 or 2 files.
      Note that in previous versions of the code the file being
      written was format 2, but it was identified as format 1.
      This error has been corrected. The default format for the file
      is 1, which is the most widely supported format.
    * Increased test coverage.
    * Fixed bug in MIDIFile for `adjust_origin` = `False`
    * Added ability to order RPC and NRPC sub-events in time, as
      a work-around for sequencers that aren't good about preserving
      event ordering for events at the time time.
    * Updates to documentation.
    * Added Travis CI to the build process. Verifying operation
      on python 2.6 - 3 development version.
    * Functions added:
        * `addTimeSignature()`
        * `addCopyright()`
        * `addText()`
        * `addKeySignature()`

Date:       23 September 2016
Version:    1.0.1
    * Minor updates to build system so that code can be hosted at
      Pypi (https://pypi.python.org/pypi/MIDIUtil/).

Date:       23 September 2016
Version:    1.0.0
    * Code ported to GitHub
    * Extensive updates to documentation, ported to Sphinx.
    * Added ability to *not* adjust the MIDIFile's time origin.
      Default behaviour is maintained, but will change in a future
      version (by default the origin is currently adjusted).
    * Changed the controller event parameter names to make them
      clearer.
    * Added support for Registered and Non-Registered parameter
      calls (``makeRPNCall`` and ``makeNRPNCall``).
    * General refactoring and clean-up.
    * Added function to select tuning program and bank.
      Some synthesizers, such as fluidsynth, require that
      uploaded tunings be explicitly assigned to a channel.
      This can be used after ``setNoteTuning`` in such a
      case.
    * Completed port to Python 3 / Unification of code base.
      Support for python < 2.6 has been dropped so that the
      Python 2 and 3 codebases could be unified.
    * Changes the way that sorting works, simplifying it and making
      it more expressive. The primary sort is on time; secondary on 
      ordinality for the event (which is user-definable, but defaults
      to an ordinality for the class); and the 
      third is the order in which the events were added.  Thus 
      is becomes easier to, say, make an RPN call, which entails 
      and ordered series of control change events all occurring at the 
      same time and of the same type.
    * Added 'annotation' as a parameter to note addition function. 
      This can be used to attach an arbitrary python object to the 
      note event. This is useful for extension development.

Date:       1 December 2013
Version:    0.89
    * Updated MIDIFile to support non-integral note values better.
    * Changed default temporal resolution to 960 ticks per beat.

examples/single-note-example.py
############################################################################
# A sample program to create a single-track MIDI file, add a note,
# and write to disk.
############################################################################

# Import the library
from midiutil import MIDIFile

# Create the MIDIFile Object
MyMIDI = MIDIFile(1)

# Add track name and tempo. The first argument to addTrackName and
# addTempo is the time to write the event.
track = 0
time = 0
MyMIDI.addTrackName(track, time, "Sample Track")
MyMIDI.addTempo(track, time, 120)

# Add a note. addNote expects the following information:
channel = 0
pitch = 60
duration = 1
volume = 100

# Now add the note.
MyMIDI.addNote(track, channel, pitch, time, duration, volume)

# And write it to disk.
with open("output.mid", 'wb') as binfile:
    MyMIDI.writeFile(binfile)


src/unittests/miditest.py
#!/usr/bin/env python
#-----------------------------------------------------------------------------
# Name:        miditest.py
# Purpose:     Unit testing harness for midiutil
#
# Author:      Mark Conway Wirt <emergentmusics) at (gmail . com>
#
# Created:     2008/04/17
# Copyright:   (c) 2009-2016, Mark Conway Wirt
# License:     Please see License.txt for the terms under which this
#              software is distributed.
#-----------------------------------------------------------------------------


from __future__ import division, print_function
import sys,  struct

import unittest
from midiutil.MidiFile import MIDIFile, writeVarLength,  \
    frequencyTransform,  returnFrequency, TICKSPERBEAT

class Decoder(object):
    '''
    An immutable comtainer for MIDI data. This is needed bcause if one indexes
    into a byte string in Python 3 one gets an ``int`` as a return.
    '''
    def __init__(self, data):
        self.data = data.decode("ISO-8859-1")
        
    def __len__(self):
        return len(self.data)
        
    def __getitem__(self, key):
        return self.data[key].encode("ISO-8859-1")
        
    def unpack_into_byte(self, key):
        return struct.unpack('>B', self[key])[0]

class TestMIDIUtils(unittest.TestCase):
    
    def testWriteVarLength(self):
        self.assertEqual(writeVarLength(0x70), [0x70])
        self.assertEqual(writeVarLength(0x80), [0x81, 0x00])
        self.assertEqual(writeVarLength(0x1FFFFF), [0xFF, 0xFF, 0x7F])
        self.assertEqual(writeVarLength(0x08000000), [0xC0, 0x80, 0x80, 0x00])
        
    def testAddNote(self):
        MyMIDI = MIDIFile(1)
        MyMIDI.addNote(0, 0, 100,0,1,100)
        self.assertEqual(MyMIDI.tracks[0].eventList[0].type, "note")
        self.assertEqual(MyMIDI.tracks[0].eventList[0].pitch, 100)
        self.assertEqual(MyMIDI.tracks[0].eventList[0].time, 0)
        self.assertEqual(MyMIDI.tracks[0].eventList[0].duration, 1)
        self.assertEqual(MyMIDI.tracks[0].eventList[0].volume, 100)
        
    def testShiftTrack(self):
        time = 1
        MyMIDI = MIDIFile(1)
        MyMIDI.addNote(0, 0, 100,time,1,100)
        self.assertEqual(MyMIDI.tracks[0].eventList[0].type, "note")
        self.assertEqual(MyMIDI.tracks[0].eventList[0].pitch, 100)
        self.assertEqual(MyMIDI.tracks[0].eventList[0].time, time)
        self.assertEqual(MyMIDI.tracks[0].eventList[0].duration, 1)
        self.assertEqual(MyMIDI.tracks[0].eventList[0].volume, 100)
        MyMIDI.shiftTracks()
        self.assertEqual(MyMIDI.tracks[0].eventList[0].time, 0)

    def testDeinterleaveNotes(self):
        MyMIDI = MIDIFile(1)
        MyMIDI.addNote(0, 0, 100, 0, 2, 100)
        MyMIDI.addNote(0, 0, 100, 1, 2, 100)
        MyMIDI.close()
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'NoteOn')
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].time,  0)
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[1].type, 'NoteOff')
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[1].time,  TICKSPERBEAT)
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[2].type, 'NoteOn')
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[2].time,  0)
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[3].type, 'NoteOff')
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[3].time,  TICKSPERBEAT * 2)
        
    def testTimeShift(self):
        
        # With one track
        MyMIDI = MIDIFile(1)
        MyMIDI.addNote(0, 0, 100, 5, 1, 100)
        MyMIDI.close()
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'NoteOn')
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].time,  0)
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[1].type, 'NoteOff')
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[1].time,  TICKSPERBEAT)
        
        # With two tracks
        MyMIDI = MIDIFile(2)
        MyMIDI.addNote(0, 0, 100, 5, 1, 100)
        MyMIDI.addNote(1, 0, 100, 6, 1, 100)
        MyMIDI.close()
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'NoteOn')
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].time,  0)
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[1].type, 'NoteOff')
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[1].time,  TICKSPERBEAT)
        self.assertEqual(MyMIDI.tracks[1].MIDIEventList[0].type, 'NoteOn')
        self.assertEqual(MyMIDI.tracks[1].MIDIEventList[0].time,  TICKSPERBEAT)
        self.assertEqual(MyMIDI.tracks[1].MIDIEventList[1].type, 'NoteOff')
        self.assertEqual(MyMIDI.tracks[1].MIDIEventList[1].time,  TICKSPERBEAT)
        
        # Negative Time
        MyMIDI = MIDIFile(1)
        MyMIDI.addNote(0, 0, 100, -5, 1, 100)
        MyMIDI.close()
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'NoteOn')
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].time,  0)
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[1].type, 'NoteOff')
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[1].time,  TICKSPERBEAT)
        
        # Negative time, two tracks
        
        MyMIDI = MIDIFile(2)
        MyMIDI.addNote(0, 0, 100, -1, 1, 100)
        MyMIDI.addNote(1, 0, 100, 0, 1, 100)
        MyMIDI.close()
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'NoteOn')
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].time,  0)
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[1].type, 'NoteOff')
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[1].time,  TICKSPERBEAT)
        self.assertEqual(MyMIDI.tracks[1].MIDIEventList[0].type, 'NoteOn')
        self.assertEqual(MyMIDI.tracks[1].MIDIEventList[0].time,  TICKSPERBEAT)
        self.assertEqual(MyMIDI.tracks[1].MIDIEventList[1].type, 'NoteOff')
        self.assertEqual(MyMIDI.tracks[1].MIDIEventList[1].time,  TICKSPERBEAT)
 
    def testFrequency(self):
        freq = frequencyTransform(8.1758)
        self.assertEqual(freq[0],  0x00)
        self.assertEqual(freq[1],  0x00)
        self.assertEqual(freq[2],  0x00)
        freq = frequencyTransform(8.66196) # 8.6620 in MIDI documentation
        self.assertEqual(freq[0],  0x01)
        self.assertEqual(freq[1],  0x00)
        self.assertEqual(freq[2],  0x00)
        freq = frequencyTransform(440.00)
        self.assertEqual(freq[0],  0x45)
        self.assertEqual(freq[1],  0x00)
        self.assertEqual(freq[2],  0x00)
        freq = frequencyTransform(440.0016)
        self.assertEqual(freq[0],  0x45)
        self.assertEqual(freq[1],  0x00)
        self.assertEqual(freq[2],  0x01)
        freq = frequencyTransform(439.9984)
        self.assertEqual(freq[0],  0x44)
        self.assertEqual(freq[1],  0x7f)
        self.assertEqual(freq[2],  0x7f)
        freq = frequencyTransform(8372.0190)
        self.assertEqual(freq[0],  0x78)
        self.assertEqual(freq[1],  0x00)
        self.assertEqual(freq[2],  0x00)
        freq = frequencyTransform(8372.062) #8372.0630 in MIDI documentation
        self.assertEqual(freq[0],  0x78)
        self.assertEqual(freq[1],  0x00)
        self.assertEqual(freq[2],  0x01)
        freq = frequencyTransform(13289.7300)
        self.assertEqual(freq[0],  0x7F)
        self.assertEqual(freq[1],  0x7F)
        self.assertEqual(freq[2],  0x7E)
        freq = frequencyTransform(12543.8760)
        self.assertEqual(freq[0],  0x7F)
        self.assertEqual(freq[1],  0x00)
        self.assertEqual(freq[2],  0x00)
        freq = frequencyTransform(8.2104) # Just plain wrong in documentation, as far as I can tell.
        #self.assertEqual(freq[0],  0x0)
        #self.assertEqual(freq[1],  0x0)
        #self.assertEqual(freq[2],  0x1)
        
        # Test the inverse
        testFreq = 15.0
        accuracy = 0.00001
        x = returnFrequency(frequencyTransform(testFreq))
        delta = abs(testFreq - x)
        self.assertEqual(delta < (accuracy*testFreq), True)
        testFreq = 200.0
        x = returnFrequency(frequencyTransform(testFreq))
        delta = abs(testFreq - x)
        self.assertEqual(delta < (accuracy*testFreq), True)
        testFreq = 400.0
        x = returnFrequency(frequencyTransform(testFreq))
        delta = abs(testFreq - x)
        self.assertEqual(delta < (accuracy*testFreq), True)
        testFreq = 440.0
        x = returnFrequency(frequencyTransform(testFreq))
        delta = abs(testFreq - x)
        self.assertEqual(delta < (accuracy*testFreq), True)
        testFreq = 1200.0
        x = returnFrequency(frequencyTransform(testFreq))
        delta = abs(testFreq - x)
        self.assertEqual(delta < (accuracy*testFreq), True)
        testFreq = 5000.0
        x = returnFrequency(frequencyTransform(testFreq))
        delta = abs(testFreq - x)
        self.assertEqual(delta < (accuracy*testFreq), True)
        testFreq = 12000.0
        x = returnFrequency(frequencyTransform(testFreq))
        delta = abs(testFreq - x)
        self.assertEqual(delta < (accuracy*testFreq), True)
        
    
    def testSysEx(self):
        #import pdb; pdb.set_trace()
        MyMIDI = MIDIFile(1)
        MyMIDI.addSysEx(0,0, 0, struct.pack('>B', 0x01))
        MyMIDI.close()
        
        data = Decoder(MyMIDI.tracks[0].MIDIdata)
        
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'SysEx')
        
        self.assertEqual(data.unpack_into_byte(0), 0x00)
        self.assertEqual(data.unpack_into_byte(1), 0xf0)
        self.assertEqual(data.unpack_into_byte(2), 3)
        self.assertEqual(data.unpack_into_byte(3), 0x00)
        self.assertEqual(data.unpack_into_byte(4), 0x01)
        self.assertEqual(data.unpack_into_byte(5), 0xf7)
        
    def testTempo(self):
        #import pdb; pdb.set_trace()
        tempo = 60
        MyMIDI = MIDIFile(1)
        MyMIDI.addTempo(0, 0, tempo)
        MyMIDI.close()
        
        data = Decoder(MyMIDI.tracks[0].MIDIdata)
        
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'Tempo')
        
        self.assertEqual(data.unpack_into_byte(0), 0x00) # time
        self.assertEqual(data.unpack_into_byte(1), 0xff) # Code
        self.assertEqual(data.unpack_into_byte(2), 0x51)
        self.assertEqual(data.unpack_into_byte(3), 0x03)
        self.assertEqual(data[4:7], struct.pack('>L', int(60000000/tempo))[1:4])
        
    def testTimeSignature(self):
        time = 0
        track = 0
        numerator = 4
        denominator = 2
        clocks_per_tick = 24
        MyMIDI = MIDIFile(1)
        MyMIDI.addTimeSignature(track, time, numerator, denominator, clocks_per_tick)
        MyMIDI.close()
        
        data = Decoder(MyMIDI.tracks[0].MIDIdata)
        
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'TimeSignature')
        
        self.assertEqual(data.unpack_into_byte(0), 0x00) # time
        self.assertEqual(data.unpack_into_byte(1), 0xFF) # Code
        self.assertEqual(data.unpack_into_byte(2), 0x58) # subcode
        self.assertEqual(data.unpack_into_byte(3), 0x04) # Data length
        self.assertEqual(data.unpack_into_byte(4), numerator)
        self.assertEqual(data.unpack_into_byte(5), denominator)
        self.assertEqual(data.unpack_into_byte(6), clocks_per_tick) # Data length
        self.assertEqual(data.unpack_into_byte(7), 0x08) # 32nd notes per quarter note
        
    def testProgramChange(self):
        #import pdb; pdb.set_trace()
        program = 10
        channel = 0
        MyMIDI = MIDIFile(1)
        MyMIDI.addProgramChange(0, channel, 0, program)
        MyMIDI.close()
        
        data = Decoder(MyMIDI.tracks[0].MIDIdata)
        
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'ProgramChange')
        self.assertEqual(data.unpack_into_byte(0), 0x00) # time
        self.assertEqual(data.unpack_into_byte(1), 0xC << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(2), program)
        
    def testTrackName(self):
        #import pdb; pdb.set_trace()
        track_name = "track"
        MyMIDI = MIDIFile(1)
        MyMIDI.addTrackName(0, 0, track_name)
        MyMIDI.close()

        data = Decoder(MyMIDI.tracks[0].MIDIdata)

        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'TrackName')
        
        self.assertEqual(data.unpack_into_byte(0), 0x00) # time
        self.assertEqual(data.unpack_into_byte(1), 0xFF) # Code
        self.assertEqual(data.unpack_into_byte(2), 0x03) # subcodes
        
    def testTuningBank(self):
        #import pdb; pdb.set_trace()
        bank = 1
        channel = 0
        MyMIDI = MIDIFile(1)
        MyMIDI.changeTuningBank(0, 0, 0, bank)
        MyMIDI.close()
        
        data = Decoder(MyMIDI.tracks[0].MIDIdata)
        
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'ControllerEvent')
        
        self.assertEqual(data.unpack_into_byte(0),  0x00)               # time
        self.assertEqual(data.unpack_into_byte(1),  0xB << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(2),  0x65)               # Controller Number
        self.assertEqual(data.unpack_into_byte(3),  0x0)                # Controller Value
        self.assertEqual(data.unpack_into_byte(4),  0x00)               # time
        self.assertEqual(data.unpack_into_byte(5),  0xB << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(6),  0x64)               # Controller Number
        self.assertEqual(data.unpack_into_byte(7),  0x4)                # Controller Value
        self.assertEqual(data.unpack_into_byte(8),  0x00)               # time
        self.assertEqual(data.unpack_into_byte(9),  0xB << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(10), 0x06)               # Bank MSB
        self.assertEqual(data.unpack_into_byte(11), 0x00)               # Value
        self.assertEqual(data.unpack_into_byte(12), 0x00)               # time
        self.assertEqual(data.unpack_into_byte(13), 0xB << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(14), 0x26)               # Bank LSB
        self.assertEqual(data.unpack_into_byte(15), bank)               # Bank value (bank number)
        
    def testTuningProgram(self):
        #import pdb; pdb.set_trace()
        program = 10
        channel = 0
        MyMIDI = MIDIFile(1)
        MyMIDI.changeTuningProgram(0, 0, 0, program)
        MyMIDI.close()

        data = Decoder(MyMIDI.tracks[0].MIDIdata)
        
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'ControllerEvent')
        
        self.assertEqual(data.unpack_into_byte(0), 0x00)               # time
        self.assertEqual(data.unpack_into_byte(1), 0xB << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(2), 0x65)               # Controller Number
        self.assertEqual(data.unpack_into_byte(3), 0x0)                # Controller Value
        self.assertEqual(data.unpack_into_byte(4), 0x00)               # time
        self.assertEqual(data.unpack_into_byte(5), 0xB << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(6), 0x64)               # Controller Number
        self.assertEqual(data.unpack_into_byte(7), 0x03)               # Controller Value
        self.assertEqual(data.unpack_into_byte(8), 0x00)               # time
        self.assertEqual(data.unpack_into_byte(9), 0xB << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(10), 0x06)              # Bank MSB
        self.assertEqual(data.unpack_into_byte(11), 0x00)              # Value
        self.assertEqual(data.unpack_into_byte(12), 0x00)              # time
        self.assertEqual(data.unpack_into_byte(13), 0xB << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(14), 0x26)               # Bank LSB
        self.assertEqual(data.unpack_into_byte(15), program)            # Bank value (bank number)
        
    def testNRPNCall(self):
        #import pdb; pdb.set_trace()
        track = 0
        time = 0
        channel = 0
        controller_msb = 1
        controller_lsb =  2
        data_msb = 3
        data_lsb = 4
        MyMIDI = MIDIFile(1)
        MyMIDI.makeNRPNCall(track, channel, time, controller_msb, controller_lsb, data_msb, data_lsb)
        MyMIDI.close()
        
        data = Decoder(MyMIDI.tracks[0].MIDIdata)
        
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'ControllerEvent')
        
        self.assertEqual(data.unpack_into_byte(0), 0x00)               # time
        self.assertEqual(data.unpack_into_byte(1), 0xB << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(2), 99)                 # Controller Number
        self.assertEqual(data.unpack_into_byte(3), controller_msb)     # Controller Value
        self.assertEqual(data.unpack_into_byte(4), 0x00)               # time
        self.assertEqual(data.unpack_into_byte(5), 0xB << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(6), 98)                 # Controller Number
        self.assertEqual(data.unpack_into_byte(7), controller_lsb)     # Controller Value
        self.assertEqual(data.unpack_into_byte(8), 0x00)               # time
        self.assertEqual(data.unpack_into_byte(9), 0xB << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(10), 0x06)              # Bank MSB
        self.assertEqual(data.unpack_into_byte(11), data_msb)          # Value
        self.assertEqual(data.unpack_into_byte(12), 0x00)              # time
        self.assertEqual(data.unpack_into_byte(13), 0xB << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(14), 0x26)               # Bank LSB
        self.assertEqual(data.unpack_into_byte(15), data_lsb) # Bank value (bank number)
        
    def testAddControllerEvent(self):
        #import pdb; pdb.set_trace()
        track = 0
        time = 0
        channel = 3
        controller_number = 1
        parameter =  2
        MyMIDI = MIDIFile(1)
        MyMIDI.addControllerEvent(track, channel, time, controller_number, parameter)
        MyMIDI.close()
        
        data = Decoder(MyMIDI.tracks[0].MIDIdata)
        
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'ControllerEvent')

        self.assertEqual(data.unpack_into_byte(0), 0x00) # time
        self.assertEqual(data.unpack_into_byte(1), 0xB << 4 | channel) # Code
        self.assertEqual(data.unpack_into_byte(2), controller_number) # Controller Number
        self.assertEqual(data.unpack_into_byte(3), parameter) # Controller Value
        
    def testNonRealTimeUniversalSysEx(self):
        code           = 1
        subcode        = 2
        payload_number = 42
        
        payload = struct.pack('>B', payload_number)

        MyMIDI = MIDIFile(1, adjust_origin=False)
        
        # Just for fun we'll use a multi-byte time
        time = 1
        time_bytes = writeVarLength(time*TICKSPERBEAT)
        MyMIDI.addUniversalSysEx(0, time, code, subcode, payload, realTime=False)
        MyMIDI.close()
        
        data = Decoder(MyMIDI.tracks[0].MIDIdata)
        
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'UniversalSysEx')
        
        self.assertEqual(data.unpack_into_byte(0), time_bytes[0]) # Time
        self.assertEqual(data.unpack_into_byte(1), time_bytes[1]) # Time
        self.assertEqual(data.unpack_into_byte(2), 0xf0) # UniversalSysEx == 0xF0
        self.assertEqual(data.unpack_into_byte(3), 5 + len(payload))    # Payload length = 5+actual pyayload
        self.assertEqual(data.unpack_into_byte(4), 0x7E) # 0x7E == non-realtime
        self.assertEqual(data.unpack_into_byte(5), 0x7F) # Sysex channel (always 0x7F)
        self.assertEqual(data.unpack_into_byte(6), code)
        self.assertEqual(data.unpack_into_byte(7), subcode)
        self.assertEqual(data.unpack_into_byte(8), payload_number) # Data
        self.assertEqual(data.unpack_into_byte(9), 0xf7) # End of message
        
    def testRealTimeUniversalSysEx(self):
        code           = 1
        subcode        = 2
        payload_number = 47
        
        payload = struct.pack('>B', payload_number)
        MyMIDI = MIDIFile(1)
        MyMIDI.addUniversalSysEx(0, 0, code, subcode, payload, realTime=True)
        MyMIDI.close()
        
        data = Decoder(MyMIDI.tracks[0].MIDIdata)
        
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'UniversalSysEx')
        
        self.assertEqual(data.unpack_into_byte(0), 0x00)
        self.assertEqual(data.unpack_into_byte(1), 0xf0)
        self.assertEqual(data.unpack_into_byte(2), 5 + len(payload))
        self.assertEqual(data.unpack_into_byte(3), 0x7F) # 0x7F == real-time
        self.assertEqual(data.unpack_into_byte(4), 0x7F)
        self.assertEqual(data.unpack_into_byte(5), code)
        self.assertEqual(data.unpack_into_byte(6), subcode)
        self.assertEqual(data.unpack_into_byte(7), payload_number)
        self.assertEqual(data.unpack_into_byte(8), 0xf7)
        
    def testTuning(self):
        MyMIDI = MIDIFile(1)
        MyMIDI.changeNoteTuning(0, [(1, 440), (2, 880)])
        MyMIDI.close()
        
        data = Decoder(MyMIDI.tracks[0].MIDIdata)
        
        self.assertEqual(MyMIDI.tracks[0].MIDIEventList[0].type, 'UniversalSysEx')
        
        self.assertEqual(data.unpack_into_byte(0), 0x00)
        self.assertEqual(data.unpack_into_byte(1), 0xf0)
        self.assertEqual(data.unpack_into_byte(2), 15)
        self.assertEqual(data.unpack_into_byte(3), 0x7F)
        self.assertEqual(data.unpack_into_byte(4), 0x7F)
        self.assertEqual(data.unpack_into_byte(5), 0x08)
        self.assertEqual(data.unpack_into_byte(6), 0x02)
        self.assertEqual(data.unpack_into_byte(7), 0x00)
        self.assertEqual(data.unpack_into_byte(8), 0x2)
        self.assertEqual(data.unpack_into_byte(9), 0x1)
        self.assertEqual(data.unpack_into_byte(10), 69)
        self.assertEqual(data.unpack_into_byte(11), 0)
        self.assertEqual(data.unpack_into_byte(12), 0)
        self.assertEqual(data.unpack_into_byte(13), 0x2)
        self.assertEqual(data.unpack_into_byte(14), 81)
        self.assertEqual(data.unpack_into_byte(15), 0)
        self.assertEqual(data.unpack_into_byte(16), 0)
        self.assertEqual(data.unpack_into_byte(17), 0xf7)
        
    def testWriteFile(self):
        # Just to make sure the stream can be written without throwing an error.
        MyMIDI = MIDIFile(1)
        MyMIDI.addNote(0, 0, 100,0,1,100)
        with open("/tmp/test.mid", "wb") as output_file:
            MyMIDI.writeFile(output_file)

def suite():
    MIDISuite = unittest.TestLoader().loadTestsFromTestCase(TestMIDIUtils)

    return MIDISuite

if __name__ == '__main__':
    print("Begining MIDIUtil Test Suite")
    MIDISuite = suite()
    unittest.TextTestRunner(verbosity=2, stream=sys.stdout).run(MIDISuite)


